File: packages/server/ExchangeGame/GameObjects/Player.cs
using ExchangeGame.Messaging;
using ExchangeGame.Repositories;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ExchangeGame
{

    public class Player: GameObject
    {
        public ICollection<Attendee> Attendees { get; private set; }

        public ICollection<Call> Calls { get; private set; }

        public Action<string> SendMessage { get; set; }

        public Player(string displayName, Action<string> sendMessage = null) : base(displayName)
        {
            _logger.LogInformation($"Player {DisplayName} is in the game");
            Attendees = ContentRepository.GetAttendees(5, this);
            Calls = new List<Call>();

            SendMessage = sendMessage ?? _defaultSendMessage;
        }

        private Action<string> _defaultSendMessage = message => { };

        public void AddCall(Call call)
        {
            Calls.Add(call);
            call.OnComplete += OnCallComplete;
        }

        private void OnCallComplete(Call call)
        {
            Calls.Remove(call);
            _logger.LogInformation($"Player {DisplayName} removing call {call.DisplayName}");
        }
    }
}

File: packages/server-v2/src/GameObjects/Player.ts
import Attendee from "./Attendee";
import Call from "./Call";
import { generateId, getAttendees } from "../utils";

export default class Player {
  public attendees: Attendee[];
  public calls: Call[];
  public sendMessage: (message: object) => void;
  public ready: boolean = false;
  public displayName: string;
  public id: number;

  constructor(displayName: string, sendMessage: (message: object) => void) {
    this.displayName = displayName;
    this.id = generateId();
    console.log(`Player ${this.displayName} ${this.id} is in the game`)
    this.attendees = getAttendees(5, this);
    this.calls = [];
    this.sendMessage = sendMessage ?? (() => undefined);
  }

  public addCall(call: Call) {
    this.calls.push(call);
    call.onComplete.push(this.onCallComplete);
  }

  private onCallComplete = (call: Call) => {
    const idx = this.calls.indexOf(call);
    if (idx !== -1) {
      this.calls.splice(idx, 1);
    }
    this.calls.splice(idx, 1);
    console.log(`Player ${this.displayName} ${this.id} removing call ${call.displayName}`)
  }
}